kernel/typing.ml

```diff
@@ -180,45 +249,64 @@ let rec pseudo_u sg (fail: int*term*term-> unit) (sigma:SS.t) : (int*term*term)
         | DB (_,_,n), t when n>=q ->
-          begin
-            let n' = n-q in
+          if sure_occur_check sg q (fun k -> k <= q || k = n) t
+          then warn()
+          else begin
             match unshift_reduce sg q t with
-            | None -> warn ()
+            | None    -> keepon()
             | Some ut ->
-               let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
-               if Subst.occurs n' t' then warn ()
-               else pseudo_u sg fail (SS.add sigma n' t') lst
+              let n' = n-q in
+              let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
+              if Subst.occurs n' t' then warn()
+              else pseudo_u sg fail (SS.add sigma n' t') lst
           end
         | t, DB (_,_,n) when n>=q ->
-          begin
-            let n' = n-q in
+          if sure_occur_check sg q (fun k -> k <= q || k = n) t
+          then warn()
+          else begin
             match unshift_reduce sg q t with
-            | None -> warn ()
+            | None    -> keepon()
             | Some ut ->
-               let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
-               if Subst.occurs n' t' then warn ()
-               else pseudo_u sg fail (SS.add sigma n' t') lst
+              let n' = n-q in
+              let t' = if Subst.occurs n' ut then ut else R.snf sg ut in
+              if Subst.occurs n' t' then warn()
+              else pseudo_u sg fail (SS.add sigma n' t') lst
           end
 
-        | App (DB (_,_,n),_,_), _  when n >= q ->
-          if R.are_convertible sg t1' t2' then keepon () else warn ()
-        | _ , App (DB (_,_,n),_,_) when n >= q ->
-          if R.are_convertible sg t1' t2' then keepon () else warn ()
-
-        | App (Const (l,cst),_,_), _ when not (Signature.is_static sg l cst) -> keepon ()
-        | _, App (Const (l,cst),_,_) when not (Signature.is_static sg l cst) -> keepon ()
+        (* f t1 ... tn    /    X t1 ... tn  =  u
+           1) Gather all free variables in t1 ... tn
+           2) Make sure u only relies on these variables
+        *)
+        | App (DB (_,_,n),a,args), t when n >= q ->
+          let occs = gather_free_vars q (a::args) in
+          if sure_occur_check sg q (fun k -> k < q && not occs.(k)) t then warn() else keepon()
+        | t, App (DB (_,_,n),a,args) when n >= q ->
+          let occs = gather_free_vars q (a::args) in
+          if sure_occur_check sg q (fun k -> k < q && not occs.(k)) t then warn() else keepon()
+        | App (Const (l,cst),a,args), t when not (Signature.is_static sg l cst) ->
+          let occs = gather_free_vars q (a::args) in
+          if sure_occur_check sg q (fun k -> k < q && not occs.(k)) t then warn() else keepon()
+        | t, App (Const (l,cst),a,args) when not (Signature.is_static sg l cst) ->
+          let occs = gather_free_vars q (a::args) in
+          if sure_occur_check sg q (fun k -> k < q && not occs.(k)) t then warn() else keepon()
 
```

Comment:
I still don't understand why don't you declare a local function to factorize the two lines that are repeated four times here

Reply:
Is it possible to write
```
        | App (DB (_,_,n),a,args), t when n >= q
        | t, App (DB (_,_,n),a,args) when n >= q
        | App (Const (l,cst),a,args), t when not (Signature.is_static sg l cst)
        | t, App (Const (l,cst),a,args) when not (Signature.is_static sg l cst) ->
          let occs = gather_free_vars q (a::args) in
          if sure_occur_check sg q (fun k -> k < q && not occs.(k)) t then warn() else keepon()
```? (because of the `when`'s I am not sure)

Reply:
No, it is not possible sadly.
@francoisthire, you suggests is to isolate the two duplicated lines and duplicate a function call instead ?
```
let case_app l cst a args t =
    let occs = gather_free_vars q (a::args) in
    if sure_occur_check sg q (fun k -> k < q && not occs.(k)) t then warn() else keepon()
in
...
match ...
...
| App (Const (l,cst),a,args), t when not (Signature.is_static sg l cst) ->
    case_app l cst a args t
| t, App (Const (l,cst),a,args) when not (Signature.is_static sg l cst) ->
    case_app l cst a args t
...
```
I'm not sure this would improve readability. A function is defined (locally) but is only called in certain cases...